// @aws-cdk/cdk/lib/cloudformation/pseudo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cloudformation_token_1 = require("./cloudformation-token");
class PseudoParameter extends cloudformation_token_1.CloudFormationToken {
    constructor(name) {
        super({ Ref: name }, name);
    }
}
exports.PseudoParameter = PseudoParameter;
class AwsAccountId extends PseudoParameter {
    constructor() {
        super('AWS::AccountId');
    }
}
exports.AwsAccountId = AwsAccountId;
class AwsDomainSuffix extends PseudoParameter {
    constructor() {
        super('AWS::DomainSuffix');
    }
}
exports.AwsDomainSuffix = AwsDomainSuffix;
class AwsURLSuffix extends PseudoParameter {
    constructor() {
        super('AWS::URLSuffix');
    }
}
exports.AwsURLSuffix = AwsURLSuffix;
class AwsNotificationARNs extends PseudoParameter {
    constructor() {
        super('AWS::NotificationARNs');
    }
}
exports.AwsNotificationARNs = AwsNotificationARNs;
class AwsNoValue extends PseudoParameter {
    constructor() {
        super('AWS::NoValue');
    }
}
exports.AwsNoValue = AwsNoValue;
class AwsPartition extends PseudoParameter {
    constructor() {
        super('AWS::Partition');
    }
}
exports.AwsPartition = AwsPartition;
class AwsRegion extends PseudoParameter {
    constructor() {
        super('AWS::Region');
    }
}
exports.AwsRegion = AwsRegion;
class AwsStackId extends PseudoParameter {
    constructor() {
        super('AWS::StackId');
    }
}
exports.AwsStackId = AwsStackId;
class AwsStackName extends PseudoParameter {
    constructor() {
        super('AWS::StackName');
    }
}
exports.AwsStackName = AwsStackName;
