// @aws-cdk/cdk/lib/cloudformation/secret.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const construct_1 = require("../core/construct");
const tokens_1 = require("../core/tokens");
const parameter_1 = require("./parameter");
/**
 * A token that represents a value that's expected to be a secret, like
 * passwords and keys.
 *
 * It is recommended to use the `SecretParameter` construct in order to import
 * secret values from the SSM Parameter Store instead of storing them in your
 * code.
 *
 * However, you can also just pass in values, like any other token: `new Secret('bla')`
 */
class Secret extends tokens_1.Token {
}
exports.Secret = Secret;
/**
 * Defines a secret value resolved from the Systems Manager (SSM) Parameter
 * Store during deployment. This is useful for referencing values that you do
 * not wish to include in your code base, such as secrets, passwords and keys.
 *
 * This construct will add a CloudFormation parameter to your template bound to
 * an SSM parameter (of type "AWS::SSM::Parameter::Value<String>"). Deployment
 * will fail if the value doesn't exist in the target environment.
 *
 * Important: For values other than secrets, prefer to use the
 * `SSMParameterProvider` which resolves SSM parameter in design-time, and
 * ensures that stack deployments are deterministic.
 */
class SecretParameter extends construct_1.Construct {
    constructor(parent, name, props) {
        super(parent, name);
        const param = new parameter_1.Parameter(this, 'Parameter', {
            type: 'AWS::SSM::Parameter::Value<String>',
            default: props.ssmParameter,
            description: props.description,
            allowedPattern: props.allowedPattern,
            allowedValues: props.allowedValues,
            constraintDescription: props.constraintDescription,
            maxLength: props.maxLength,
            minLength: props.minLength,
            noEcho: true,
        });
        this.value = new Secret(param.ref);
    }
}
exports.SecretParameter = SecretParameter;
