// @aws-cdk/cdk/test/cloudformation/test.output.js
"use strict";
const lib_1 = require("../../lib");
module.exports = {
    'outputs can be added to the stack'(test) {
        const stack = new lib_1.Stack();
        const res = new lib_1.Resource(stack, 'MyResource', { type: 'R' });
        const ref = new lib_1.Ref(res);
        new lib_1.Output(stack, 'MyOutput', {
            export: 'ExportName',
            value: ref,
            description: 'Output properties'
        });
        test.deepEqual(stack.toCloudFormation(), { Resources: { MyResource: { Type: 'R' } },
            Outputs: { MyOutput: { Description: 'Output properties',
                    Export: { Name: 'ExportName' },
                    Value: { Ref: 'MyResource' } } } });
        test.done();
    },
    'outputs cannot be referenced'(test) {
        const stack = new lib_1.Stack();
        const output = new lib_1.Output(stack, 'MyOutput', { description: 'My Output' });
        test.throws(() => output.ref);
        test.done();
    },
    'outputs have a default unique export name'(test) {
        const stack = new lib_1.Stack(undefined, 'MyStack');
        const output = new lib_1.Output(stack, 'MyOutput');
        const child = new lib_1.Construct(stack, 'MyConstruct');
        const output2 = new lib_1.Output(child, 'MyOutput2');
        test.equal(output.export, 'MyStack:MyOutput');
        test.equal(output2.export, 'MyStack:MyConstructMyOutput255322D15');
        test.done();
    },
    'disableExport can be used to disable the auto-export behavior'(test) {
        const stack = new lib_1.Stack();
        const output = new lib_1.Output(stack, 'MyOutput', { disableExport: true });
        test.equal(output.export, null);
        // cannot specify `export` and `disableExport` at the same time.
        test.throws(() => new lib_1.Output(stack, 'YourOutput', {
            disableExport: true,
            export: 'bla'
        }), /Cannot set `disableExport` and specify an export name/);
        test.done();
    },
    'is stack name is undefined, we will only use the logical ID for the export name'(test) {
        const stack = new lib_1.Stack();
        const output = new lib_1.Output(stack, 'MyOutput');
        test.equal(output.export, 'MyOutput');
        test.done();
    },
    'makeImportValue can be used to create an Fn::ImportValue from an output'(test) {
        const stack = new lib_1.Stack(undefined, 'MyStack');
        const output = new lib_1.Output(stack, 'MyOutput');
        test.deepEqual(lib_1.resolve(output.makeImportValue()), { 'Fn::ImportValue': 'MyStack:MyOutput' });
        test.done();
    }
};
