// @aws-cdk/cdk/test/cloudformation/test.secret.js
"use strict";
const lib_1 = require("../../lib");
module.exports = {
    'Secret is merely a token'(test) {
        const foo = new lib_1.Secret('Foo');
        const bar = new lib_1.Secret(() => 'Bar');
        test.deepEqual(lib_1.resolve(foo), 'Foo');
        test.deepEqual(lib_1.resolve(bar), 'Bar');
        test.done();
    },
    'SecretParameter can be used to define values resolved from SSM parameter store during deployment'(test) {
        const stack = new lib_1.Stack();
        const mySecret = new lib_1.SecretParameter(stack, 'MySecret', { ssmParameter: '/my/secret/param' });
        new lib_1.SecretParameter(stack, 'Boom', {
            ssmParameter: 'Boom',
            description: 'description',
            constraintDescription: 'constraintDescription',
            minLength: -100,
            maxLength: 2000,
            allowedPattern: 'allowed-pattern',
            allowedValues: ['allowed', 'values'],
        });
        test.deepEqual(stack.toCloudFormation(), { Parameters: { MySecretParameterBB81DE58: { Type: 'AWS::SSM::Parameter::Value<String>',
                    Default: '/my/secret/param',
                    NoEcho: true },
                BoomParameterB3EB3942: { Type: 'AWS::SSM::Parameter::Value<String>',
                    Default: 'Boom',
                    AllowedPattern: 'allowed-pattern',
                    AllowedValues: ['allowed', 'values'],
                    ConstraintDescription: 'constraintDescription',
                    Description: 'description',
                    MaxLength: 2000,
                    MinLength: -100,
                    NoEcho: true } } });
        // value resolves to a "Ref"
        test.deepEqual(lib_1.resolve(mySecret.value), { Ref: 'MySecretParameterBB81DE58' });
        test.done();
    }
};
